### PR TITLE
[Object Spilling] Introduce SpillWorker & RestoreWorker Pool to avoid IO worker deadlock.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Previously, there could be deadlock when many spilling/restore objects happen concurrently. Imagine all IO workers are occupied to restore objects while the object store is about to OOM. The put request will fail, and the plasma store needs to spill objects, but all IO workers are already occupied by restore ops.

This PR fixes this issue by having a separate pool for spill/restore workers. It basically removes the concept of IO workers (it will be just an internal concept within worker pool abstraction) and introduces SpillWorker and RestoreWorker.

This PR doesn't actually allow users to set a separate pool size for spill/restore workers. I think having the same size for both pools is fine for now, and implementing it in the future upon demand is trivial. 

It also introduces tests for IO worker methods within worker pool (which we didn't have for some reason). 

## Related issue number

https://github.com/ray-project/ray/issues/11789

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -629,15 +633,18 @@ cdef void restore_spilled_objects_handler(
...
+                external_storage.restore_spilled_objects(urls)
         except Exception:
             exception_str = (
                 "An unexpected internal error occurred while the IO worker "
                 "was restoring spilled objects.")
             logger.exception(exception_str)
             ray.utils.push_error_to_driver(
                 ray.worker.global_worker,
-                "io_worker_retore_spilled_objects_error",
+                "retore_spilled_objects_error",
```
Comment (ericl) [inline python/ray/_raylet.pyx:647]: retore => restore

File: python/ray/ray_constants.py [hunk 2]
```diff
@@ -159,7 +159,16 @@ def to_memory_units(memory_bytes, round_up):
 PROCESS_TYPE_GCS_SERVER = "gcs_server"
 
 WORKER_PROCESS_TYPE_IDLE_WORKER = "ray::IDLE"
-WORKER_PROCESS_TYPE_IO_WORKER = "ray::IOWorker"
+WORKER_PROCESS_TYPE_SPILL_WORKER_PREFIX = "ray::SpillWorker"
+WORKER_PROCESS_TYPE_RESTORE_WORKER_PREFIX = "ray::RestoreWorker"
+WORKER_PROCESS_TYPE_SPILL_WORKER_IDLE = (
```
Comment (ericl) [inline python/ray/ray_constants.py:164]: Could we call it IDLE_SpillWorker since process names are usually truncated to the first few characters in `top`?

**Review by ericl — APPROVED** [review]

File: python/ray/ray_constants.py (hunk 2 above)
Comment (rkooo567) [reply python/ray/ray_constants.py:164]: In this case, will it look like

IDLE: `ray::IDLE_SpillWorker`
spill: `ray::SPILL_SpillWorker`?

Comment (rkooo567) [reply python/ray/ray_constants.py:164]: Or do you completely want to remove ray::?

Comment (rkooo567) [reply python/ray/ray_constants.py:164]: Changed to 

IDLE: ray::IDLE_SpillWorker
spill: ray::SPILL_SpillWorker

IDLE: ray::IDLE_RestoreWorker
restore: ray::RESTORE_RestoreWorker

**Pull request merged**